quaireaux/math/src/sha256.cairo

```diff
@@ -0,0 +1,321 @@
+    data.append(downcast(shr(u64_data_length, 8) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 0) & U8_MAX).unwrap());
+
+    let u32_data_length = 16 * ((data.len() - 1) / 64 + 1);
+    let mut data = from_u8Array_to_u32Array(ref data, u32_data_length);
+    let mut h = get_h();
+    let mut k = get_k();
+    h = sha256_inner(ref data, 0, ref k, h);
+
+    from_u32Array_to_u8Array(ref h, 8)
+}
+
+fn from_u32Array_to_u8Array(ref data: Array<u32>, i: usize) -> Array<u8> {
+    quaireaux_utils::check_gas();
+    if i <= 0 {
+        return ArrayTrait::new();
+    }
+    let mut result = from_u32Array_to_u8Array(ref data, i - 1);
+    result.append(downcast(shr(upcast(*data[i - 1]), 24) & U8_MAX).unwrap());
+    result.append(downcast(shr(upcast(*data[i - 1]), 16) & U8_MAX).unwrap());
+    result.append(downcast(shr(upcast(*data[i - 1]), 8) & U8_MAX).unwrap());
+    result.append(downcast(shr(upcast(*data[i - 1]), 0) & U8_MAX).unwrap());
+    result
+}
+
+fn copy_array(ref from: Array<u32>, i: usize) -> Array<u32> {
+    quaireaux_utils::check_gas();
+    if i <= 0 {
+        return ArrayTrait::new();
+    }
+    let mut result = copy_array(ref from, i - 1);
+    result.append(*from[i - 1]);
+    result
+}
+
+fn sha256_inner(
+    ref data: Array<u32>, i: usize, ref k: Array<u32>, mut h: Array<u32>
+) -> Array<u32> {
+    quaireaux_utils::check_gas();
+    if 16 * i >= data.len() {
+        return h;
+    }
+    let mut w = create_w(ref data, i, 16);
+    create_message_schedule(ref w, 16);
+    let mut h2 = copy_array(ref h, h.len());
+    let mut h2 = compression(ref w, 0, ref k, h2);
+    let mut t = ArrayTrait::new();
+    t.append(u32_wrapping_add(*h[0], *h2[0]));
+    t.append(u32_wrapping_add(*h[1], *h2[1]));
+    t.append(u32_wrapping_add(*h[2], *h2[2]));
+    t.append(u32_wrapping_add(*h[3], *h2[3]));
+    t.append(u32_wrapping_add(*h[4], *h2[4]));
+    t.append(u32_wrapping_add(*h[5], *h2[5]));
+    t.append(u32_wrapping_add(*h[6], *h2[6]));
+    t.append(u32_wrapping_add(*h[7], *h2[7]));
+    h = t;
+    sha256_inner(ref data, i + 1, ref k, h)
+}
+
+fn compression(ref w: Array<u32>, i: usize, ref k: Array<u32>, mut h: Array<u32>) -> Array<u32> {
+    quaireaux_utils::check_gas();
+    if i >= 64 {
+        return h;
+    }
+    let s1 = bsig1(*h[4]);
+    let ch = ch(*h[4], *h[5], *h[6]);
+    let temp1 = u32_wrapping_add(
+        u32_wrapping_add(u32_wrapping_add(u32_wrapping_add(*h[7], s1), ch), *k[i]), *w[i]
+    );
+    let s0 = bsig0(*h[0]);
+    let maj = maj(*h[0], *h[1], *h[2]);
+    let temp2 = u32_wrapping_add(s0, maj);
+    let mut t = ArrayTrait::new();
+    t.append(u32_wrapping_add(temp1, temp2));
+    t.append(*h[0]);
+    t.append(*h[1]);
+    t.append(*h[2]);
+    t.append(u32_wrapping_add(*h[3], temp1));
+    t.append(*h[4]);
+    t.append(*h[5]);
+    t.append(*h[6]);
+    h = t;
+    h = compression(ref w, i + 1, ref k, h);
+    h
```

Comment:
in compression, no need to pass ref w it can be just w
plus, those two last lines could be just one:
`compression(w, i + 1, ref k, h)`